src/ff1.rs

```diff
@@ -90,6 +98,54 @@ impl RadixOps<u16> for u16 {
     }
 }
 
+/// A radix 2^i for i in [1..16). It does not use floating-point arithmetic.
+#[derive(Debug, PartialEq)]
+pub struct PowerTwoRadix {
+    radix: u16,
+    log_radix: u8,
+}
+
+impl PowerTwoRadix {
+    pub fn from(radix: u16) -> Result<Self, ()> {
+        let mut tmp = radix;
+        let mut log_radix = 0;
+        let mut found = false;
+        for i in 0..16 {
+            if tmp & 1 != 0 {
+                // Only a single bit should be set
+                if found {
+                    return Err(());
+                }
+                log_radix = i;
+                found = true;
+            }
+            tmp >>= 1;
+        }
+        match log_radix {
+            0 => Err(()),
+            _ => Ok(PowerTwoRadix { radix, log_radix }),
+        }
+    }
+}
+
+impl RadixOps for PowerTwoRadix {
+    fn check_in_range(&self, n: u32) -> bool {
+        n >> self.log_radix == 0
+    }
+
+    fn calculate_b(&self, v: usize) -> usize {
+        ((v * self.log_radix as usize) + 7) / 8
+    }
+
+    fn to_biguint(&self) -> BigUint {
+        BigUint::from(self.radix)
+    }
+
+    fn to_u32(&self) -> u32 {
+        self.radix as u32
+    }
```

Comment:
I would have just used ``u32`` for the radix.

Reply:
Yeah, now that I've done the large refactor I think it's fine to do that, so I'll make those changes in a follow-up PR (along with the then-necessary constraining of the flexible radix to 2^16).